### PR TITLE
macOS small extensions and fixes 

### PR DESCRIPTION
**Pull request by csomor**

### CODE REVIEW

File: src/osx/carbon/dataobj.cpp [hunk 1]
```diff
@@ -219,8 +219,23 @@ void wxDataFormat::SetId( NativeFormat format )
...
+    {
+        wxUniChar c = *it;
+        if ( !( c >= 'A' && c <='Z') && !( c >= 'a' && c <='z') && !( c >= '0' && c <='9') &&
+            c != '.' && c !='-' )
+            *it= '-';
+    }
+
+    // if not having a '.' we add our own reverse-DNS
+    if ( utiString.Index('.') == wxNOT_FOUND)
+        utiString = "org.wxwidgets."+utiString;
```
Comment (vadz) [inline src/osx/carbon/dataobj.cpp:237]: This is trivial, but I'd use `std::string`-compatible methods, i.e.
```suggestion
    if ( utiString.find('.') == wxString::npos )
        utiString.insert(0, "org.wxwidgets.");
```

File: src/generic/treectlg.cpp [hunk 2]
```diff
@@ -3010,6 +3010,10 @@ void wxGenericTreeCtrl::OnSize( wxSizeEvent &event )
 void wxGenericTreeCtrl::OnPaint( wxPaintEvent &WXUNUSED(event) )
 {
     wxPaintDC dc(this);
+    #ifdef __WXOSX__
+    // at least on OSX otherwise the non-painted area remains transparent
```
Comment (vadz) [inline src/generic/treectlg.cpp:3014]: This would seem to indicate that wxOSX doesn't implement the contract of wx API correctly: it's supposed to erase background with the default `wxBG_STYLE_ERASE` background style, so that `wxEVT_PAINT` handler could paint only the "foreground" over it.

Note that in this particular case we actually could set the background style to `wxBG_STYLE_PAINT` and call `dc.Clear()` here too, but the current code should still work, so it looks like something could be wrong in `wxWindowMac::MacDoRedraw()`, could you please check it?

File: src/osx/cocoa/listbox.mm [hunk 3]
```diff
@@ -375,7 +381,7 @@ wxAlignment WXUNUSED(just), int defaultWidth)
         [col1 setMinWidth:10];
         // temporary hack, because I cannot get the automatic column resizing
         // to work properly
-        [col1 setWidth:1000];
+        // [col1 setWidth:1000];
```
Comment (vadz) [inline src/osx/cocoa/listbox.mm:384]: Could we get rid of this now?

File: src/osx/cocoa/overlay.mm [hunk 4]
```diff
@@ -78,6 +78,10 @@ - (instancetype)initWithContentRect:(NSRect)contentRect styleMask:(NSUInteger)st
     return self;
 }
 
+- (void)dealloc
+{
+   [super dealloc];
+}
```
Comment (vadz) [inline src/osx/cocoa/overlay.mm:84]: Just for my own education, could you please explain why is this necessary? I.e. wouldn't superclass dealloc be used by default anyhow?

File: src/osx/carbon/font.cpp [hunk 5]
```diff
@@ -819,6 +819,19 @@ void wxNativeFontInfo::InitFromFontDescriptor(CTFontDescriptorRef desc)
...
+
+void wxNativeFontInfo::AdjustPostScriptName()
+{
+    if ( wxCheckOsVersion( 10, 16) )
+    {
+        // the PostScript names reported in macOS start with a dot for System Fonts, this has to be corrected
+        // otherwise round-trips are not possible, resulting in a Times Fallback, therefore we replace these with
+        // their official PostScript Name
+        if ( m_postScriptName.StartsWith(".SFNS") )
+            m_postScriptName.Replace(".SFNS","SFProDisplay");
```
Comment (vadz) [inline src/osx/carbon/font.cpp:833]: In theory, this could replace more than one occurrence of `.SFNS`, should it occur more than once. You could just pass `false` as the last parameter here or do something like
```suggestion
        wxString rest;
        if ( m_postScriptName.StartsWith(".SFNS", &rest) )
            m_postScriptName = "SFProDisplay" + rest;
```

File: include/wx/fontutil.h [hunk 6]
```diff
@@ -179,6 +179,9 @@ class WXDLLIMPEXP_CORE wxNativeFontInfo
     
     void RealizeResource() const;
 private:
+    // the PostScript names reported in macOS start with a dot for System Fonts, this has to be corrected
+    // otherwise round-trips are not possible, resulting in a Times Fallback
```
Comment (vadz) [inline include/wx/fontutil.h:183]: Just trying to understand: why isn't round trip possible for these fonts? Is it a wx problem or does underlying API handle leading periods specially or something else?

File: include/wx/fontutil.h [hunk 7]
```diff
@@ -179,6 +179,9 @@ class WXDLLIMPEXP_CORE wxNativeFontInfo
...
+    void AdjustPostScriptName();
```
Comment (vadz) [inline include/wx/fontutil.h:184]: This class doesn't really follow this convention (probably because it predates it), but normally this should be
```suggestion
    void OSXAdjustPostScriptName();
```

File: src/osx/carbon/font.cpp [hunk 8]
```diff
@@ -819,6 +819,19 @@ void wxNativeFontInfo::InitFromFontDescriptor(CTFontDescriptorRef desc)
 
     wxCFTypeRef(CTFontDescriptorCopyAttribute(m_descriptor, kCTFontFamilyNameAttribute)).GetValue(m_familyName);
     wxCFTypeRef(CTFontDescriptorCopyAttribute(m_descriptor, kCTFontNameAttribute)).GetValue(m_postScriptName);
+    AdjustPostScriptName();
+}
+
+void wxNativeFontInfo::AdjustPostScriptName()
+{
+    if ( wxCheckOsVersion( 10, 16) )
```
Comment (vadz) [inline src/osx/carbon/font.cpp:827]: Interesting that [Trac comment](https://trac.wxwidgets.org/ticket/18878#comment:9) claims that the same problem also happens with 10.15. Could we also have `.SFNS`-prefixed fonts there? Or is it due to something completely different on that system?

**Comment by vadz** [comment]
@csomor Can/should I cherry-pick commits that don't raise any questions from this PR? This would basically leave c40813a (which I really think should be fixed at `wxWindow` level) and 43dc1ce (which I think should be fixed in MSW/GTK instead).

File: src/osx/carbon/dataobj.cpp (hunk 1 above)
Comment (csomor) [reply src/osx/carbon/dataobj.cpp:237]: ok, I agree, thanks

File: src/osx/cocoa/overlay.mm (hunk 4 above)
Comment (csomor) [reply src/osx/cocoa/overlay.mm:84]: I got a warning without that, that my property would not be properly released, apparently the code for this only gets generated if there's a destructor. I couldn't find any notice about this behavior anywhere, but I followed clang's suggestion here ..

File: include/wx/fontutil.h (hunk 6 above)
Comment (csomor) [reply include/wx/fontutil.h:183]: macOS issued a warning about this in earlier systems, but in 11.00 makes it fail, defaulting to Times ... that way at least it looks better, although it's not a perfect fix, they want us to use the numeric constants for OS-Fonts

File: src/osx/cocoa/listbox.mm (hunk 3 above)
Comment (csomor) [reply src/osx/cocoa/listbox.mm:384]: unfortunately this still is not properly resolved for the case when a horizontal scrollbar is wanted, then the column's size is only initially calculated correctly, if it later gets longer strings, the content will be eclipsed with '..' instead of increasing column size

https://trac.wxwidgets.org/ticket/18860

File: include/wx/fontutil.h (hunk 7 above)
Comment (csomor) [reply include/wx/fontutil.h:184]: I'm fine with that change

File: src/osx/carbon/font.cpp (hunk 8 above)
Comment (csomor) [reply src/osx/carbon/font.cpp:827]: these names have been around for long times, but they didn't cause problems before. I wrote in that ticket that it must be a slightly different situation which I'd like to understand first before blindly applying this to 10.15 as well ..

**Comment by csomor** [comment]
> @csomor Can/should I cherry-pick commits that don't raise any questions from this PR? This would basically leave [c40813a](https://github.com/wxWidgets/wxWidgets/commit/c40813a8fa4eab393db7b66ad938766a25d21d12) (which I really think should be fixed at `wxWindow` level) and [43dc1ce](https://github.com/wxWidgets/wxWidgets/commit/43dc1cefa44492acb2de27f72a3adc54aa540470) (which I think should be fixed in MSW/GTK instead).

please go ahead, in the wxChoice case I'm not really sure, I think since we already documented:
"The label of the item or an empty string if the position was invalid."
I'd rather apply my fix"
I'll look at the remaining issue with the tree control. On mac does it does not have a background color set, it is on default, which on mac where many native controls are transparent seems reasonable, I guess just setting this explicitly will make things work.

**Comment by vslavik** [comment]
The b494443 commit shouldn't be cherry-picked just yet because it isn't fully functional, see #2034.

**Comment by vadz** [comment]
Sorry, I have no idea what to do with this one, there are really too many different things here and too many questions. In commit order:

- 43dc1cefa4 (behave according to msw, gtk and the docs, don’t assert in debug with invalid index): shouldn't be applied, I will modify the other ports and add a test for this instead, see https://trac.wxwidgets.org/ticket/18852
- 4ae329472c (make private data name conform to UTI rules): might not be needed any longer and conflicts with 14fb1c5fe1 (wrapping private types properly in UTI, 2020-11-29) that I think has replaced it.
- 8cc694b919 (feeding back standard commands like copy, paste into wxMenItem if possible): from reading https://trac.wxwidgets.org/ticket/18738 it's not really clear if this is needed, i.e. whether it's supposed to be merged or was just for testing, could you please check if it is (still) needed? If so, it should be cherry-picked (I can do this, of course).
- 1432df0edf (moving wxNSApplication to header): not clear what is this for, `wxNSApplication` doesn't seem to be used anywhere. OTOH `transformToForegroundApplication` probably should be declared, seeing as it's implemented.
- c40813a8fa (makes sure the entire area is cleared properly on macOS): I think this one isn't needed as it was another workaround for https://trac.wxwidgets.org/ticket/18940 which was fixed differently.
- b66b752295 (adding a layout method to trigger column width recalculation): I believe this one is not needed after 701334a3ba (Fix wxListBox horizontal scrollbar updating in wxOSX, 2020-12-05).
- 747a78afd9 (fixing return value in case of success …): This one is clearly needed, I've cherry-picked it now.
- f420e39719 (fixing memory leak): I still don't really understand this one, but I've cherry-picked it too.
- b494443989 (fixing system font variants under macOS Big Sur): This seems to have been replaced by b03fa9c37f (Fix UI font serialization on macOS 11, 2020-08-27), but I'm really not sure. Could you please check?
- The remaining commits are fixups for the previous ones, so I don't discuss them separately.

Please let me know if any the remaining ones need to be looked at/applied, otherwise I think we can close this one. TIA!

**Comment by csomor** [comment]
@vadz thanks for looking at these

> * [4ae3294](https://github.com/wxWidgets/wxWidgets/commit/4ae329472c2c668b4793f2ea388ab2e23069484d) (make private data name conform to UTI rules): might not be needed any longer and conflicts with [14fb1c5](https://github.com/wxWidgets/wxWidgets/commit/14fb1c5fe1d0206360a7cd28dd197793ab6e22d1) (wrapping private types properly in UTI, 2020-11-29) that I think has replaced it.

partly, the iterating through the string, replacing illegal chars is missing, I can that directly to master though ...

> * [8cc694b](https://github.com/wxWidgets/wxWidgets/commit/8cc694b9199b718b351b3b1346f19654eb05516d) (feeding back standard commands like copy, paste into wxMenItem if possible): from reading https://trac.wxwidgets.org/ticket/18738 it's not really clear if this is needed, i.e. whether it's supposed to be merged or was just for testing, could you please check if it is (still) needed? If so, it should be cherry-picked (I can do this, of course).

yes, I'd like to have this

> * [1432df0](https://github.com/wxWidgets/wxWidgets/commit/1432df0edfa641376a2dc97939b82b3169f0e0bb) (moving wxNSApplication to header): not clear what is this for, `wxNSApplication` doesn't seem to be used anywhere. OTOH `transformToForegroundApplication` probably should be declared, seeing as it's implemented.

IIRC that was a wish from a 3rd party app developer, to be able to call this, therefore I've moved it to the private header

> * [f420e39](https://github.com/wxWidgets/wxWidgets/commit/f420e3971933354ba5afe97c6ad5d5af83929c86) (fixing memory leak): I still don't really understand this one, but I've cherry-picked it too.

setContentView owns the view, so we must autorelease it before, otherwise it will never be released

> * [b494443](https://github.com/wxWidgets/wxWidgets/commit/b494443989397846f3b30df54b3a1d08e64c91d5) (fixing system font variants under macOS Big Sur): This seems to have been replaced by [b03fa9c](https://github.com/wxWidgets/wxWidgets/commit/b03fa9c37f97230130b37fdf99eb8517235c32c2) (Fix UI font serialization on macOS 11, 2020-08-27), but I'm really not sure. Could you please check?

yes, it's not needed anymore

> Please let me know if any the remaining ones need to be looked at/applied, otherwise I think we can close this one. TIA!

I'll do so, thanks!

**Comment by vadz** [comment]
Thanks, I've cherry-picked 8cc694b (not sure if this is enough to fix the ticket or not?) and will let you apply any other changes to the master/make new PRs for them.

Thanks again!
